File: 975. Odd Even Jump/main.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

class Solution {
public:
    int oddEvenJumps(vector<int>& A) {
        int N = A.size();
        int res = 0;
        map<int,int> m;
        vector<bool> dp_odd(N, false);
        vector<bool> dp_even(N, false);

        m[A[N-1]] = N-1;
        dp_odd[A.size()-1] = true;
        dp_even[A.size()-1] = true;
        res++;

        for(int i = A.size()-2; i>=0; i--) {
            auto it_low = m.lower_bound(A[i]);
            auto it_up  = m.upper_bound(A[i]);
            if(it_low != m.end()) {
                dp_odd[i] = dp_even[it_low->second];
            }
            if(it_up != m.begin()) {
                dp_even[i] = dp_odd[(--it_up)->second];
            }

            if(dp_odd[i]) res++;
            m[A[i]] = i;
        }

        return res;
    }
};

int main()
{
    vector<int> v1 = {10,13,12,14,15};
    vector<int> v2 = {2,3,1,1,4};
    vector<int> v3 = {5,1,3,4,2};

    Solution sol;
    cout<<sol.oddEvenJumps(v1)<<endl;
    cout<<sol.oddEvenJumps(v2)<<endl;
    cout<<sol.oddEvenJumps(v3)<<endl;
    return 0;
}

File: 975. Odd Even Jump/main.py
from typing import List
import bisect

class Solution:
    # TLE
    def oddEvenJumps_v2(self, A: List[int]) -> int:
        N = len(A)
        dp_odd  = [False] * (N-1) + [True]
        dp_even = [False] * (N-1) + [True]
        
        vals = [[A[N-1],N-1]]
        res = 1
        for i in range(N-2, -1, -1):
            keys = [r[0] for r in vals]
            lo = bisect.bisect_left(keys, A[i])
            hi = bisect.bisect_right(keys, A[i])
            if lo < len(vals):
                dp_odd[i] = dp_even[vals[lo][1]]
                if dp_odd[i]:
                    res += 1

            if hi > 0:
                dp_even[i] = dp_odd[vals[hi-1][1]]

            if lo < len(vals) and vals[lo][0] == A[i]:
                vals[lo][1] = i
            else:
                vals.insert(lo, [A[i],i])
        return res

    def oddEvenJumps(self, A: List[int]) -> int:

        N = len(A)
        dp_odd  = [0] * N
        dp_even = [0] * N

        Aasc = sorted([(v,i) for (i,v) in enumerate(A)])
        st = []
        for j in range(len(Aasc)):
            while st and st[-1] < Aasc[j][1]:
                dp_odd[st[-1]] = Aasc[j][1]
                st.pop()
            st.append(Aasc[j][1])
        
        Adec = sorted([(-v,i) for (i,v) in enumerate(A)])
        st = []
        for j in range(len(Adec)):
            while st and st[-1] < Adec[j][1]:
                dp_even[st[-1]] = Adec[j][1]
                st.pop()
            st.append(Adec[j][1])

        odd  = [0] * (N-1) + [1]
        even = [0] * (N-1) + [1]
        for j in range(N-1)[::-1]:
            odd[j]  = even[dp_odd[j]]
            even[j] = odd[dp_even[j]]

        return sum(odd)

if __name__ == "__main__":
    sol = Solution()
    print(sol.oddEvenJumps([10,13,12,14,15]))
    print(sol.oddEvenJumps([2,3,1,1,4]))
    print(sol.oddEvenJumps([5,1,3,4,2]))